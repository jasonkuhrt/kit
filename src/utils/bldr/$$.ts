export * from './constructor.js'
export * from './mutable.js'
export * from './state.js'
export * from './types.js'
