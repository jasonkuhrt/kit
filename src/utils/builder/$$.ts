export * from './constructor.js'
export * from './state.js'
export * from './types.js'
