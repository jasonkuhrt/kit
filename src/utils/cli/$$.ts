export * from './cli.js'
export * as FlagName from './flag-name/$$.js'
