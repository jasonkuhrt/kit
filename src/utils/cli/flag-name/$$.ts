export * as Analyzer from './analyzer.js'
export * as Checks from './checks.js'
export { FlagName } from './name.js'

// Type-level exports
export type { Analyze, Errors, SomeLimits } from './analyzer.types.js'
