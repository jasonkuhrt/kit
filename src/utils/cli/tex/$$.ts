/**
 * Terminal Text Layout Engine
 *
 * Build rich CLI text layouts with a chainable API. Create bordered blocks, tables,
 * and lists with automatic width calculation, padding, and visual alignment.
 *
 * @category CLI/Text Rendering
 *
 * @example
 * ```typescript
 * import { Tex } from '@wollybeard/kit'
 *
 * // Simple text block
 * const output = Tex.Tex()
 *   .text('Hello World')
 *   .render()
 *
 * // Bordered block with padding
 * Tex.Tex()
 *   .block({
 *     border: { top: '-', bottom: '-', left: '|', right: '|' },
 *     padding: { left: 2, right: 2 }
 *   }, 'Important Message')
 *   .render()
 *
 * // Table with headers
 * Tex.Tex()
 *   .table(($) => $
 *     .headers(['Name', 'Age', 'City'])
 *     .row('Alice', '30', 'NYC')
 *     .row('Bob', '25', 'LA')
 *   )
 *   .render()
 *
 * // Nested blocks
 * Tex.Tex()
 *   .block(($) => $
 *     .text('Parent Block')
 *     .block({ padding: { left: 2 } }, 'Nested Block')
 *   )
 *   .render()
 * ```
 */

export type { BlockBuilder } from './chain/block.js'
export { block } from './chain/block.js'
export type { Builder } from './chain/helpers.js'
export type { RootBuilder } from './chain/root.js'

/**
 * Create a root text layout builder.
 *
 * @category CLI/Text Rendering
 * @param parameters - Optional block parameters for the root container, including terminalWidth
 * @returns A root builder with chainable methods and `.render()` to generate output
 *
 * @example
 * ```typescript
 * // Start building and render
 * const output = Tex.Tex()
 *   .text('Hello')
 *   .text('World')
 *   .render()
 *
 * // With explicit terminal width (useful for testing or fixed-width output)
 * Tex.Tex({ terminalWidth: 80 })
 *   .text('Content')
 *   .render()
 *
 * // With other root-level configuration
 * Tex.Tex({ terminalWidth: 120, padding: { left: 2 } })
 *   .text('Content')
 *   .render()
 * ```
 */
export { createRootBuilder as Tex } from './chain/root.js'

/**
 * Render a builder to a string without calling `.render()` on the builder itself.
 * Useful for rendering builders created in utility functions.
 *
 * @category CLI/Text Rendering
 * @param builder - Any tex builder (root, block, table, or list)
 * @returns The rendered text output
 *
 * @example
 * ```typescript
 * const builder = Tex.Tex().text('Hello')
 * const output = Tex.render(builder)
 * // Same as: builder.render()
 * ```
 */
export { render } from './chain/root.js'

export type { TableBuilder } from './chain/table.js'
export type { BlockParameters } from './nodes/block.js'

/**
 * Block node class. Can be used to create blocks directly instead of via builder methods.
 *
 * @category CLI/Text Rendering
 *
 * @example
 * ```typescript
 * // Usually you use the builder API:
 * Tex.Tex().block('content')
 *
 * // But you can create blocks directly:
 * const myBlock = new Tex.Block({ border: { top: '-' } }, 'content')
 * Tex.Tex().block(myBlock).render()
 * ```
 */
export { Block } from './nodes/block.js'
