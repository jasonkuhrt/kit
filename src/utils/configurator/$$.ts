export { Configurator } from './configurator.js'
export type * from './configurator.js'

// Re-export commonly used types from Configurator namespace at the top level for easier access
import type { Configurator as C } from './configurator.js'
export type ApplyConfiguratorInputResolver$Func<
  $Configurator extends C,
  $Current extends $Configurator['normalizedIncremental'],
  $Input extends $Configurator['input'],
> = C.ApplyConfiguratorInputResolver$Func<$Configurator, $Current, $Input>
export type BuilderProviderCallback<$C extends C> = C.BuilderProviderCallback<$C>
export type Builder<$C extends C> = C.Builder<$C>
export type Configuration = C.Configuration
export type DataInput<$C extends C = C> = C.DataInput<$C>
export type InferParameters<$C extends C> = C.InferParameters<$C>
