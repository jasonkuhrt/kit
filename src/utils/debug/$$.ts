export * from './debug.js'
export * from './dump.js'
export * from './trace.js'
