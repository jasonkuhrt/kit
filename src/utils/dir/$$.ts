export type { Dir } from './dir.js'
export { create, createTemp, createTempUnsafe } from './dir.js'

export type { DirChain } from './chain.js'
export { chain, withChaining } from './chain.js'

export type { Operation, SpecBuilder } from './spec.js'
export { spec } from './spec.js'

export { executeOperations } from './operations.js'
