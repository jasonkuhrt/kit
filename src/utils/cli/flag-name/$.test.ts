import { Test } from '#test'
import { Ts } from '#ts'
import { Schema as S } from 'effect'
import { test } from 'vitest'
import type { Analyze, Errors } from './$$.js'
import { Checks, FlagName } from './$$.js'

// =============================================================================
// Type-Level Parsing Tests (ported from mixed.spec.ts)
// =============================================================================

test('type-level parsing: basic cases', () => {
  type _ = Ts.Test.Cases<
    // Short flag only
    Ts.Test.bid<
      { expression: string; canonical: 'v'; long: null; short: 'v'; aliases: { short: []; long: [] } },
      Analyze<'-v'>
    >,
    // Short with one alias
    Ts.Test.bid<
      { expression: string; canonical: 'v'; long: null; short: 'v'; aliases: { short: ['x']; long: [] } },
      Analyze<'-v -x'>
    >,
    // Short with multiple aliases
    Ts.Test.bid<
      { expression: string; canonical: 'v'; long: null; short: 'v'; aliases: { short: ['x', 'y']; long: [] } },
      Analyze<'-v -x -y'>
    >,
    // Long only
    Ts.Test.bid<
      { expression: string; canonical: 'vv'; long: 'vv'; short: null; aliases: { short: []; long: ['xx'] } },
      Analyze<'--vv --xx'>
    >,
    // Long with multiple aliases
    Ts.Test.bid<
      { expression: string; canonical: 'vv'; long: 'vv'; short: null; aliases: { short: []; long: ['xx', 'yy'] } },
      Analyze<'--vv --xx --yy'>
    >,
    // Mixed short and long
    Ts.Test.bid<
      { expression: string; canonical: 'vv'; long: 'vv'; short: 'v'; aliases: { short: ['x']; long: ['xx'] } },
      Analyze<'-v --vv -x --xx'>
    >,
    // No prefix (inferred by length)
    Ts.Test.bid<
      { expression: string; canonical: 'vv'; long: 'vv'; short: 'v'; aliases: { short: ['x']; long: ['xx'] } },
      Analyze<'v vv x xx'>
    >,
    // Mixed prefix and no prefix
    Ts.Test.bid<
      { expression: string; canonical: 'vv'; long: 'vv'; short: 'v'; aliases: { short: ['x']; long: ['xx'] } },
      Analyze<'v --vv x xx'>
    >
  >
})

// =============================================================================
// Type-Level Parsing Tests (ported from other.spec.ts)
// =============================================================================

test('type-level parsing: case handling', () => {
  type _ = Ts.Test.Cases<
    // Kebab case to camelCase
    Ts.Test.bid<
      { expression: string; canonical: 'filePath'; long: 'filePath'; short: null; aliases: { short: []; long: [] } },
      Analyze<'--file-path'>
    >
  >
})

test('type-level parsing: whitespace handling', () => {
  interface SomeLong {
    expression: string
    canonical: 'version'
    long: 'version'
    short: null
    aliases: { short: []; long: [] }
  }

  type _1 = Ts.Test.Cases<
    Ts.Test.bid<SomeLong, Analyze<'--version'>>,
    Ts.Test.bid<SomeLong, Analyze<' --version'>>,
    Ts.Test.bid<SomeLong, Analyze<'  --version '>>,
    Ts.Test.bid<SomeLong, Analyze<'  --version  '>>,
    Ts.Test.bid<SomeLong, Analyze<' --version  '>>,
    Ts.Test.bid<SomeLong, Analyze<'version  '>>
  >

  interface SomeShort {
    expression: string
    canonical: 'v'
    long: null
    short: 'v'
    aliases: { short: []; long: [] }
  }

  type _2 = Ts.Test.Cases<
    Ts.Test.bid<SomeShort, Analyze<'-v'>>,
    Ts.Test.bid<SomeShort, Analyze<' -v'>>,
    Ts.Test.bid<SomeShort, Analyze<' -v '>>,
    Ts.Test.bid<SomeShort, Analyze<' -v  '>>,
    Ts.Test.bid<SomeShort, Analyze<'  -v '>>,
    Ts.Test.bid<SomeShort, Analyze<'  -v  '>>,
    Ts.Test.bid<SomeShort, Analyze<'-v  '>>,
    Ts.Test.bid<SomeShort, Analyze<'-v '>>,
    Ts.Test.bid<SomeShort, Analyze<'v '>>,
    Ts.Test.bid<SomeShort, Analyze<'v'>>
  >

  interface SomeLongShort {
    expression: string
    canonical: 'version'
    long: 'version'
    short: 'v'
    aliases: { short: []; long: [] }
  }

  type _3 = Ts.Test.Cases<
    Ts.Test.bid<SomeLongShort, Analyze<'--version -v'>>,
    Ts.Test.bid<SomeLongShort, Analyze<' --version -v'>>,
    Ts.Test.bid<SomeLongShort, Analyze<' --version -v '>>,
    Ts.Test.bid<SomeLongShort, Analyze<'  --version -v  '>>,
    Ts.Test.bid<SomeLongShort, Analyze<'  --version -v '>>,
    Ts.Test.bid<SomeLongShort, Analyze<'-v --version'>>,
    Ts.Test.bid<SomeLongShort, Analyze<' -v --version'>>,
    Ts.Test.bid<SomeLongShort, Analyze<'  -v --version'>>,
    Ts.Test.bid<SomeLongShort, Analyze<'  -v --version '>>,
    Ts.Test.bid<SomeLongShort, Analyze<'  -v --version  '>>,
    Ts.Test.bid<SomeLongShort, Analyze<'  -v  --version  '>>,
    Ts.Test.bid<SomeLongShort, Analyze<'  -v   --version '>>,
    Ts.Test.bid<SomeLongShort, Analyze<'  v   version '>>,
    Ts.Test.bid<SomeLongShort, Analyze<'v version'>>
  >

  interface SomeLongCamelCase {
    expression: string
    canonical: 'fooBar'
    long: 'fooBar'
    short: null
    aliases: { short: []; long: [] }
  }

  type _4 = Ts.Test.Cases<
    Ts.Test.bid<SomeLongCamelCase, Analyze<'--fooBar'>>,
    Ts.Test.bid<SomeLongCamelCase, Analyze<'--foo-bar'>>
  >
})

// =============================================================================
// Type-Level Validation Tests (ported from errors.spec.ts)
// =============================================================================

test('type-level validation: empty input', () => {
  type _ = Ts.Test.Cases<
    Ts.Test.bid<Errors.Empty, Analyze<''>>,
    Ts.Test.bid<Errors.Empty, Analyze<' '>>
  >
})

test('type-level validation: reserved names', () => {
  type _ = Ts.Test.Cases<
    // Short flag reserved
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'a'>>,
      Analyze<'-a', { reservedNames: 'a'; usedNames: undefined }>
    >,
    // Long flag reserved
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'abc'>>,
      Analyze<'--abc', { reservedNames: 'abc'; usedNames: undefined }>
    >,
    // Case normalization (kebab vs camel)
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'foo-bar'>>,
      Analyze<'--foo-bar', { reservedNames: 'fooBar'; usedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'fooBar'>>,
      Analyze<'--fooBar', { reservedNames: 'foo-bar'; usedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'foo-bar'>>,
      Analyze<'foo-bar', { reservedNames: 'fooBar'; usedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'fooBar'>>,
      Analyze<'fooBar', { reservedNames: 'foo-bar'; usedNames: undefined }>
    >,
    // Aliases
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'foo-bar'>>,
      Analyze<'--foo --foo-bar', { reservedNames: 'fooBar'; usedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'fooBar'>>,
      Analyze<'--foo --fooBar', { reservedNames: 'foo-bar'; usedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'foo-bar'>>,
      Analyze<'foo foo-bar', { reservedNames: 'fooBar'; usedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.Reserved<'fooBar'>>,
      Analyze<'foo fooBar', { reservedNames: 'foo-bar'; usedNames: undefined }>
    >
  >
})

test('type-level validation: already taken names', () => {
  type _ = Ts.Test.Cases<
    // Short flag already taken
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'a'>>,
      Analyze<'-a', { usedNames: 'a'; reservedNames: undefined }>
    >,
    // Long flag already taken
    Ts.Test.bid<
      'Error(s):\nThe name "abc" cannot be used because it is already used for another flag.',
      Analyze<'--abc', { usedNames: 'abc'; reservedNames: undefined }>
    >,
    // Case normalization
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'fooBar'>>,
      Analyze<'--fooBar', { usedNames: 'foo-bar'; reservedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'foo-bar'>>,
      Analyze<'--foo-bar', { usedNames: 'fooBar'; reservedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'fooBar'>>,
      Analyze<'fooBar', { usedNames: 'foo-bar'; reservedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'foo-bar'>>,
      Analyze<'foo-bar', { usedNames: 'fooBar'; reservedNames: undefined }>
    >,
    // Aliases
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'fooBar'>>,
      Analyze<'--foo --fooBar', { usedNames: 'foo-bar'; reservedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'foo-bar'>>,
      Analyze<'--foo --foo-bar', { usedNames: 'fooBar'; reservedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'fooBar'>>,
      Analyze<'foo fooBar', { usedNames: 'foo-bar'; reservedNames: undefined }>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AlreadyTaken<'foo-bar'>>,
      Analyze<'foo foo-bar', { usedNames: 'fooBar'; reservedNames: undefined }>
    >
  >
})

test('type-level validation: length errors', () => {
  type _ = Ts.Test.Cases<
    // Long flag too short
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.LongTooShort<'v'>>,
      Analyze<'--v'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.LongTooShort<'v'>>,
      Analyze<'--ver --v'>
    >,
    // Short flag too long
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.ShortTooLong<'vv'>>,
      Analyze<'-vv'>
    >
  >
})

test('type-level validation: duplicate aliases', () => {
  type _ = Ts.Test.Cases<
    // Exact duplicates
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'vv'>>,
      Analyze<'--vv --vv'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'v-v'>>,
      Analyze<'--v-v --v-v'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'v'>>,
      Analyze<'-v -v'>
    >,
    // Case variations (kebab vs camel)
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'foo-bar'>>,
      Analyze<'--fooBar --foo-bar'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'fooBar'>>,
      Analyze<'--foo-bar --fooBar'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'foo-bar'>>,
      Analyze<'fooBar foo-bar'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'fooBar'>>,
      Analyze<'foo-bar fooBar'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'fooBar'>>,
      Analyze<'foo-bar --fooBar'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'fooBar'>>,
      Analyze<'--foo-bar fooBar'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'foo-bar'>>,
      Analyze<'fooBar --foo-bar'>
    >,
    Ts.Test.bid<
      Checks.Messages.WithHeader<Checks.Messages.AliasDuplicate<'foo-bar'>>,
      Analyze<'--fooBar foo-bar'>
    >
  >
})

// =============================================================================
// Type-Level Tests for FlagName.fromString
// =============================================================================

test('type-level: fromString return types', () => {
  // Valid inputs - type inference works
  Ts.Test.bid<Analyze<'-v'>>()(FlagName.fromString('-v'))
  Ts.Test.bid<Analyze<'--verbose'>>()(FlagName.fromString('--verbose'))
  Ts.Test.bid<Analyze<'-v --verbose'>>()(FlagName.fromString('-v --verbose'))
  Ts.Test.bid<Analyze<'--foo-bar'>>()(FlagName.fromString('--foo-bar'))
  Ts.Test.bid<Analyze<'-v --verbose -x --extra'>>()(FlagName.fromString('-v --verbose -x --extra'))

  // Invalid inputs - input guard rejects with Ts.StaticError
  // These are type-only tests - they verify compile-time errors without executing
  if (false as boolean) {
    // @ts-expect-error - Empty input
    FlagName.fromString('')
  }
})

// =============================================================================
// Runtime Parsing Tests
// =============================================================================

// Runtime decoder without type-level guards
const decodeString = S.decodeSync(FlagName.String)

// dprint-ignore
Test.on(decodeString)
  .describeInputs('success', [
    ['-v'],
    ['--verbose'],
    ['-v --verbose'],
    ['--foo-bar'],
    ['-v --verbose -x --extra'],
    ['v verbose x extra'],
  ])
  .describeInputs('error', [
    [''],
    [' '],
    ['--v'],
    ['-vv'],
    ['--foo --foo'],
    ['-v -v'],
  ])
  .test()
