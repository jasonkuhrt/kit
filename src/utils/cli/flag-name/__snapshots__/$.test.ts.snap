// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`error > fn(' ') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
 
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ You must specify at least one name for your flag.
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fn('') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS

╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ You must specify at least one name for your flag.
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fn('--foo --foo') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
--foo --foo
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ Duplicate alias: "foo"
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fn('--v') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
--v
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ Long flag must be two or more characters: '--v'
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fn('-v -v') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-v -v
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ Duplicate alias: "v"
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fn('-vv') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-vv
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ Short flag must be exactly one character: '-vv'
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fromString(' ') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
 
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ You must specify at least one name for your flag.
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fromString('') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS

╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ You must specify at least one name for your flag.
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fromString('--foo --foo') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
--foo --foo
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ Duplicate alias: "foo"
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fromString('--v') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
--v
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ Long flag must be two or more characters: '--v'
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fromString('-v -v') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-v -v
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ Duplicate alias: "v"
╚══════════════════════════════════════════════════╝"
`;

exports[`error > fromString('-vv') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-vv
╠══════════════════════════════════════════════════╣ THEN THROWS PARSEERROR
ParseError: (string <-> (FlagName (Encoded side) <-> FlagName))
└─ Transformation process failure
   └─ Short flag must be exactly one character: '-vv'
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fn('--foo-bar') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
--foo-bar
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'fooBar',
  short: null,
  long: 'fooBar',
  aliases: {
    short: [],
    long: []
  },
  expression: '--foo-bar'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fn('--verbose') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
--verbose
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'verbose',
  short: null,
  long: 'verbose',
  aliases: {
    short: [],
    long: []
  },
  expression: '--verbose'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fn('-v --verbose -x --extra') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-v --verbose -x --extra
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'verbose',
  short: 'v',
  long: 'verbose',
  aliases: {
    short: [ 'x' ],
    long: [ 'extra' ]
  },
  expression: '-v --verbose -x --extra'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fn('-v --verbose') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-v --verbose
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'verbose',
  short: 'v',
  long: 'verbose',
  aliases: {
    short: [],
    long: []
  },
  expression: '-v --verbose'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fn('-v') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-v
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'v',
  short: 'v',
  long: null,
  aliases: {
    short: [],
    long: []
  },
  expression: '-v'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fn('v verbose x extra') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
v verbose x extra
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'verbose',
  short: 'v',
  long: 'verbose',
  aliases: {
    short: [ 'x' ],
    long: [ 'extra' ]
  },
  expression: 'v verbose x extra'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fromString('--foo-bar') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
--foo-bar
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'fooBar',
  short: null,
  long: 'fooBar',
  aliases: {
    short: [],
    long: []
  },
  expression: '--foo-bar'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fromString('--verbose') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
--verbose
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'verbose',
  short: null,
  long: 'verbose',
  aliases: {
    short: [],
    long: []
  },
  expression: '--verbose'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fromString('-v --verbose -x --extra') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-v --verbose -x --extra
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'verbose',
  short: 'v',
  long: 'verbose',
  aliases: {
    short: [ 'x' ],
    long: [ 'extra' ]
  },
  expression: '-v --verbose -x --extra'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fromString('-v --verbose') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-v --verbose
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'verbose',
  short: 'v',
  long: 'verbose',
  aliases: {
    short: [],
    long: []
  },
  expression: '-v --verbose'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fromString('-v') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-v
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'v',
  short: 'v',
  long: null,
  aliases: {
    short: [],
    long: []
  },
  expression: '-v'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`success > fromString('v verbose x extra') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
v verbose x extra
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
FlagName {
  canonical: 'verbose',
  short: 'v',
  long: 'verbose',
  aliases: {
    short: [ 'x' ],
    long: [ 'extra' ]
  },
  expression: 'v verbose x extra'
}
╚══════════════════════════════════════════════════╝"
`;
